openjdk/mmtkBarrierSetAssembler_x86.hpp

```diff
@@ -4,11 +4,40 @@
 #include "asm/macroAssembler.hpp"
 #include "gc/shared/barrierSetAssembler.hpp"
 
+class MMTkBarrierSetC1;
+class MMTkC1BarrierStub;
+class LIR_Assembler;
+class StubAssembler;
+
 class MMTkBarrierSetAssembler: public BarrierSetAssembler {
+  friend class MMTkBarrierSetC1;
+
+protected:
+  /// Full pre-barrier
+  virtual void object_reference_write_pre(MacroAssembler* masm, DecoratorSet decorators, Address dst, Register val, Register tmp1, Register tmp2) const {}
+  /// Full post-barrier
+  virtual void object_reference_write_post(MacroAssembler* masm, DecoratorSet decorators, Address dst, Register val, Register tmp1, Register tmp2) const {}
+
+  /// Barrier elision test
+  virtual bool can_remove_barrier(DecoratorSet decorators, Register val, bool skip_const_null) const {
+    bool in_heap = (decorators & IN_HEAP) != 0;
+    bool as_normal = (decorators & AS_NORMAL) != 0;
+    assert((decorators & IS_DEST_UNINITIALIZED) == 0, "unsupported");
+    return !in_heap || (skip_const_null && val == noreg);
+  }
+
+  /// Generate C1 write barrier slow-call assembly code
+  virtual void generate_c1_write_barrier_runtime_stub(StubAssembler* sasm) const;
```

Comment:
What is the difference between `generate_c1_write_barrier_runtime_stub` and `gen_c1_generic_write_barrier_stub`? They seem to generate similar code that calls the slow path, but used in different cases. You should consider renaming them, as the difference in their names does not tell what they are.

Reply:
`generate_c1_write_barrier_runtime_stub` will generate a small assembly function that calls MMTk barrier slow-calls. It's called a "runtime stub".

`gen_c1_generic_write_barrier_stub` will generate a piece of code in the barrier to jump to the above assembly function.

So in general, to do a slow-path barrier call in C1, it will jump to the "runtime stub" first, then call to the mmtk API.

I directly copied this code from G1, and currently I'm still not sure why jumping to the "runtime stub" first is necessary. Probably in the future I'll see if I can remove this "runtime stub".

Reply:
I see. `CodeStub` is a piece of out-of-line code (a new function) that is put at the end of a function. There is only one copy of that for each function, and for places you need it, you just generate some code to call it.

I suggest renaming `gen_c1_generic_write_barrier_stub` to `generate_c1_write_barrier_stub_call`, or something like that. It is not really generating a stub.

Reply:
Yes makes sense. Renamed.